File: assets/samples/bulk.whois/perl/bulk_whois.pl
#!/usr/bin/perl
use Package::Alias
  http => 'HTTP::Request',
  ua   => 'LWP::UserAgent',
  js   => 'JSON';

my $url = 'https://www.whoisxmlapi.com/BulkWhoisLookup/bulkServices';
my ( $apiKey, $dom, $fm ) =
  ( 'Bulk Whois API Key', ['google.com'], 'json' );
#
sub post_data {
    my ( $uri, $dt, $bs ) =
      ( @_, { apiKey => $apiKey, outputFormat => $fm } );
    my ( $ua, $req ) = ( ua->new(), http->new( 'POST', $url . $uri ) );
    $req->content( encode_json( { %$bs, %$dt } ) );
    return js->new->decode( $ua->request($req)->content );
}    #
my ( $res, $i ) =
  ( post_data( '/bulkWhois', { domains => $dom } ), 'requestId' );
$res->{messageCode}
  && ( $res->{messageCode} != 200 )
  && die $res->{messageCode};
my $arg = { $i => $res->{$i}, startIndex => @$dom + 1, maxRecords => 0 };
while ( post_data( '/getRecords', $arg )->{recordsLeft} > 0 ) { sleep(3); }
( $arg->{startIndex}, $arg->{maxRecords} ) = ( 1, @$dom + 0 );
print js->new->pretty->encode( post_data( '/getRecords', $arg ) );

File: assets/samples/bulk.whois/ruby/bulk_whois.rb
require 'json'
require 'net/https'
require 'uri'

csv_filename, domains = 'whois_records.csv', %w[google.com whoisxmlapi.com]

def api(path, data)
  url = 'https://www.whoisxmlapi.com/BulkWhoisLookup/bulkServices'
  b ={ apiKey:'Bulk Whois API Key', outputFormat:'json' }
  uri = URI.parse(url + path)
  http, http.use_ssl = Net::HTTP.new(uri.host, uri.port), true
  req, req.body = Net::HTTP::Post.new(uri.request_uri), b.merge(data).to_json
  http.request(req).body
end
# This will save whois record info for all domains as #{csv_filename}
id = JSON.parse(api('/bulkWhois', domains: domains))['requestId']
data = { requestId: id, startIndex: domains.count + 1, maxRecords: 0 }
loop do
  break if JSON.parse(api('/getRecords', data))['recordsLeft'].zero?
  sleep 3
end
File.write(csv_filename, api('/download', requestId: id, startIndex: 1))
